### PR TITLE
Callback execution with mutex synchronization

### PR DESCRIPTION
**Pull request by zhuravljov**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | no

It can be able to execute callback code with mutex synchronization. For example:

```php
$result = $mutex->sync($mutexName, 10, function () {
    // business logic execution with synchronization
    return $result;
}));
```

### CODE REVIEW

**Comment by zhuravljov** [comment]
@samdark I'm not sure how tests should cover it rightly. The tests are run in one thread mode. The feature is designed to sync code execution in multithread mode, when monopoly execution is very important.

There is chance to get order like:

- process 1 step 1
- process 2 step 1
- process 1 step 2
- process 2 step 2

But the method ensure order like:

- process 1 step 1
- process 1 step 2
- process 2 step 1
- process 2 step 2

**Comment by samdark** [comment]
Right. These cases are really hard to test :(

**Comment by rugabarbo** [comment]
https://github.com/yiisoft/yii2/blob/master/tests/framework/db/mysql/connection/DeadLockTest.php#L32

Can it help?

File: framework/mutex/Mutex.php [hunk 1]
```diff
@@ -97,6 +103,40 @@ public function release($name)
...
+            try {
+                $result = call_user_func($callback, $this);
+            } catch (\Exception $e) {
+                $this->release($name);
+                throw $e;
+            } catch (\Throwable $e) {
+                $this->release($name);
+                throw $e;
+            }
+            // TODO: use "finally" keyword when miminum required PHP version is >= 5.5
```
Comment (SilverFire) [inline framework/mutex/Mutex.php]: This PR contains a new feature, so will go to 2.1 branch. You can already use `finally` keyword

File: framework/mutex/Mutex.php [hunk 2]
```diff
@@ -97,6 +103,40 @@ public function release($name)
...
+     * Executes callback with mutex synchronization.
+     *
+     * @param string $name of the lock to be acquired. Must be unique.
+     * @param int $timeout time (in seconds) to wait for lock to be released.
+     * @param callable $callback a valid PHP callback that performs the job. Accepts mutex instance as parameter.
+     * @param bool $throw whether to throw an exception when the lock is not acquired.
+     * @return mixed result of callback function, or null when the lock is not acquired.
+     * @throws \Exception|\Throwable if there is any exception thrown from the callback.
+     * @throws SyncException when the lock is not acquired.
+     * @since (unknown yet)
```
Comment (SilverFire) [inline framework/mutex/Mutex.php]: 2.1

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/mutex/Mutex.php (hunk 1 above)
Comment (zhuravljov) [reply framework/mutex/Mutex.php]: Done

File: framework/mutex/Mutex.php (hunk 2 above)
Comment (zhuravljov) [reply framework/mutex/Mutex.php]: Done

File: framework/mutex/Mutex.php [hunk 3]
```diff
@@ -97,6 +103,33 @@ public function release($name)
...
+    /**
+     * Executes callback with mutex synchronization.
+     *
+     * @param string $name of the lock to be acquired. Must be unique.
+     * @param int $timeout time (in seconds) to wait for lock to be released.
+     * @param callable $callback a valid PHP callback that performs the job. Accepts mutex instance as parameter.
+     * @param bool $throw whether to throw an exception when the lock is not acquired.
+     * @return mixed result of callback function, or null when the lock is not acquired.
+     * @throws SyncException when the lock is not acquired.
+     * @since 2.1
```
Comment (rugabarbo) [inline framework/mutex/Mutex.php]: If so, you must create PR for 2.1 branch, **not for master**.

**Comment by rugabarbo** [comment]
And yet... what about tests?

File: framework/mutex/Mutex.php (hunk 3 above)
Comment (zhuravljov) [reply framework/mutex/Mutex.php]: @rugabarbo Mutex class is not changed in 2.1. The PR can be merged into 2.1 without conflicts.

**Comment by zhuravljov** [comment]
> And yet... what about tests?

It is simple wrapper over:

```php
if ($mutex->acquire($lockName, $timeout)) {
    //...
    $mutex->release($lockName);
}
```

If acquire/release methods are tested and work correctly, sync method will work also.

**Comment by zhuravljov** [comment]
Added three test cases. They don't test multithread, but cover each method work mode.

**Comment by samdark** [comment]
PHP 5.4 fails. The rest seems to be good.

**Comment by samdark** [comment]
Nope. PHP 5.4 doesn't have `finally`...

**Comment by samdark** [comment]
Oh, just noticed that the pull request should've been made against 2.1 branch, not master. @zhuravljov would you please fix it? Thanks.

**Comment by rugabarbo** [comment]
Yes, I already pointed to this: 
https://github.com/yiisoft/yii2/pull/15985#discussion_r178522255

**Comment by samdark** [comment]
OK. I've changed base and I think 2.1 should be merged into the pull request first...

**Comment by zhuravljov** [comment]
@samdark may be there is sense to close the PR, and open new one, that will be based on 2.1?

**Comment by samdark** [comment]
Yes.

**Pull request closed without merging**